src/lib.rs

```diff
@@ -0,0 +1,560 @@
+                if self.nodes[parent_index].left_child == Some(index) {
+                    self.nodes[parent_index].left_child = Some(left_index);
+                } else {
+                    self.nodes[parent_index].right_child = Some(left_index);
+                }
+            } else {
+                // change the root
+                self.root = Some(left_index);
+            }
+            self.nodes[index].height =
+                self.compute_height(self.nodes[index].left_child,
+                                    self.nodes[index].right_child);
+            self.nodes[left_index].height =
+                self.compute_height(self.nodes[left_index].left_child,
+                                    self.nodes[left_index].right_child);
+            self.nodes[left_index].parent = self.nodes[index].parent;
+            self.nodes[index].parent = Some(left_index);
+            self.update_heights(index);
+        }
+    }
+
+    fn right_left_rotation(&mut self, index: usize) {
+        if let Some(right_index) = self.nodes[index].right_child {
+            self.right_rotation(right_index);
+            self.left_rotation(index);
+        }
+    }
+
+    fn left_right_rotation(&mut self, index: usize) {
+        if let Some(left_index) = self.nodes[index].left_child {
+            self.left_rotation(left_index);
+            self.right_rotation(index);
+        }
+    }
+
+    fn rebalance(&mut self, index: usize) {
+        let balance_factor = self.balance_factor(&self.nodes[index]);
+        if balance_factor == -2 {
+            // right-heavy tree
+            let right_index = self.nodes[index].right_child.unwrap();
+            if self.balance_factor(&self.nodes[right_index]) == -1 {
+                self.left_rotation(index);
+            } else {
+                self.right_left_rotation(index);
+            }
+        } else if balance_factor == 2 {
+            // left-heavy
+            let left_index = self.nodes[index].left_child.unwrap();
+            if self.balance_factor(&self.nodes[left_index]) == 1 {
+                self.right_rotation(index);
+            } else {
+                self.left_right_rotation(index);
+            }
+        } else if balance_factor < -2 || balance_factor > 2 {
+            panic!("Invalid balance factor for {}: {}", index, balance_factor);
+        }
+    }
+
+    fn is_balanced(&self, index: usize) -> bool {
+        let balance_factor = self.balance_factor(&self.nodes[index]);
+        balance_factor >= -1 && balance_factor <= 1
+    }
+
+    /// Return an iterator over the values in this tree.
+    ///
+    /// # Example
+    /// ```
+    /// use avl_tree::SearchTree;
+    /// let mut tree: SearchTree<u64> = SearchTree::new();
+    /// tree.insert(5);
+    /// tree.insert(2);
+    /// tree.insert(1);
+    /// assert!(tree.iter().zip(vec![1, 2, 5]).all(|(x, y)| *x == y));
+    /// ```
+    pub fn iter(&'a self) -> SearchTreeIter<'a, T> {
+        let mut start_node = None;
```

Comment:
I feel like this initialisation code should be encapsulated inside the iterator's `::new()`

Reply:
It just walks down the left side of the tree to find the first node that should fall out of the iterator, right?

Reply:
Hm, I didn't implement it like that because it felt a bit awkward to pass both the root node and the tree `Vec` to `SearchTreeIter::new`, but I guess it makes makes sense to do it that way